src/xen/xen.c

```diff
@@ -0,0 +1,847 @@
+    if (gt->entry_heap == INVALID_ADDRESS) {
+        msg_err("failed to allocate grant table occupancy heap\n");
+        goto fail_dealloc_table;
+    }
+
+    /* We have this feature on the platforms we care about now, but we
+       can add support for the other case if need be. */
+    if (!xen_feature_supported(XENFEAT_auto_translated_physmap)) {
+        msg_err("auto translated physmap feature required\n");
+        goto fail_dealloc_heap;
+    }
+
+    /* Other implementations add a page at a time as required. For
+       simplicity, we're going to add them all on initialization. */
+    for (int p = 0; p < qs.max_nr_frames; p++) {
+        /* We know at the moment that backed is physically contiguous,
+           but best not to not assume in case this changes. */
+        u64 phys = physical_from_virtual(((void *)gt->table) + p * PAGESIZE);
+        struct xen_add_to_physmap xatp;
+        xatp.domid = DOMID_SELF;
+        xatp.space = XENMAPSPACE_grant_table;
+        xatp.idx = p;
+        xatp.gpfn = phys >> PAGELOG;
+        xen_debug("grant page %d, gpfn %x", xatp.idx, xatp.gpfn);
+        rv = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
+        if (rv < 0) {
+            msg_err("failed to add grant table page (rv %d)\n", rv);
+            goto fail_dealloc_heap;
+        }
+    }
+    return true;
+  fail_dealloc_heap:
+    destroy_heap(gt->entry_heap);
+  fail_dealloc_table:
+    deallocate(heap_backed(kh), gt->table, qs.max_nr_frames * PAGESIZE);
+    return false;
+}
+
+/* returns 0 on alloc fail */
+/* optimization: create a gntref free list / fifo of sorts */
+grant_ref_t xen_grant_page_access(u16 domid, u64 phys, boolean readonly)
+{
+    struct gtab *gt = &xen_info.gtab;
+    u64 ref = allocate_u64(gt->entry_heap, 1);
+    if (ref == INVALID_PHYSICAL)
+        return 0;
+    gt->table[ref].domid = domid;
+    gt->table[ref].frame = phys >> PAGELOG;
+    write_barrier();
+    gt->table[ref].flags = GTF_permit_access | (readonly ? GTF_readonly : 0);
+    return (grant_ref_t)ref;
+}
+
+void xen_revoke_page_access(grant_ref_t ref)
+{
+    assert(ref > 8 && ref != -1);
+    xen_info.gtab.table[ref].flags = 0;
+    memory_barrier();
+    deallocate_u64(xen_info.gtab.entry_heap, ref, 1);
+}
+
+/* Reportedly, Xen timers can fire up to 100us early. */
+#define XEN_TIMER_SLOP_NS 100000
+static CLOSURE_0_1(xen_runloop_timer, void, timestamp);
+static void xen_runloop_timer(timestamp duration)
+{
+    u64 n = pvclock_now_ns();
+    u64 expiry = n + MAX(nsec_from_timestamp(duration), XEN_TIMER_SLOP_NS);
+
+    rprintf("%s: now %T, expiry %T\n", __func__, nanoseconds(n), nanoseconds(expiry));
+    int rv = HYPERVISOR_set_timer_op(expiry);
+    if (rv != 0) {
+        msg_err("failed; rv %d\n", rv);
+    }
+}
+
+static CLOSURE_0_0(xen_runloop_timer_handler, void);
+static void xen_runloop_timer_handler(void)
+{
+    rprintf("%s: now %T\n", __func__, nanoseconds(pvclock_now_ns()));
```

Comment:
How often does this event channel fire? Should this rprintf be a debug?

Reply:
The event channel is fired by Xen when a timer expires. That should be a debug, but the code is still in flux and presently is not in use (see PR description).